pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+//! # Auction Pallet
+//!
+//! This pallet provides functionality for creating, canceling and biding auction for non-fungible
+//! tokens (NFTs). It supports role-based access control and integrates with a role management
+//! system.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, BlockNumberProvider, Saturating},
+	DispatchResult, Percent,
+};
+use sp_std::{fmt::Debug, prelude::*, vec::Vec};
+
+use primitives::{AuctionId, NftId, NftState, Role};
+use traits::{BasicAccounting, GetTreasury, ManageAuctions, ManageNfts, ManageRoles};
+
+//mod mock;
+//mod test
+
+pub mod weights;
+pub use module::*;
+pub use weights::*;
+
+//#[cfg(feature = "runtime-benchmarks")]
+//mod benchmarking;
+
+/// Represents Auction data including its reserve price, expiry block, current price and bider.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct AuctionData<T: Config> {
+	pub nft_id: NftId,
+	pub owner: T::AccountId,
```

Comment:
The owner of the NFT can be lookedup via the NFT trait. Not need to duplicate storage here